### DISCUSSION TITLE
Failed to get XSRF Token

### DISCUSSION THREAD

**Discussion by Robertsmania** [Bukk94/KickLib#1] [Q&A] [answered]
Thank you for providing this, its a huge help.

I find that every time I call kickApi.Messages.SendMessageAsync() there console output shows:
`Failed to get XSRF Token
`

Poking around I find that the output is coming from PuppeteerExtensions.GetXsrfTokenAsync()

When I call kickApi.AuthenticateAsync() - that does run through the puppeteer extensions and gets the XSRF token on the first attempt.

Then when I call SendMessagesAsync() it looks like it usually fails on the first attempt, prints out the message but then does get the token on a subsequent attempt.

Is that expected?  Is there a way I should be storing and re-using the token I got in the initial AuthenticateAsync() or otherwise provide credentials so it can use the token we already have (it looks like the token stays the same on each call?).

In spite of the output and multiple attempts to get the token, the SendMessagesAsync() function does in fact work and the message appears in the Kick chat.

**Comment by Bukk94** [comment]
It's not expected but I noticed exactly the same behavior. For some reason first attempt always fail (that's why I put retry there). When I was poking around, storing XSRF token didn't help as re-using that token caused some issues.

I can try investigate why it's doing that or if there is way around it, but I cannot guarantee I'll be able to fix it. I should find some way, because every re-try on that function just slows down overall responsiveness, so even though it will send the message in Chat eventually, it takes like 2 seconds.

**Reply by Robertsmania** [reply]
Okay, thanks for the prompt reply and its good to hear its not just user error on my part.  

For now I've added some extra Console.WriteLine statements in GetXsrfTokenAsync() to try and get a sense of what its doing over time.  

Its very early days in my project and I dont expect to have anything really ready for use for a bit.  Your stuff has been a huge help to get started.

**Reply by Bukk94** [reply]
Ok I think I got it. Xsrf token should be valid per session so I store it instead of retrieving it each time. And if the request fails on token mismatch, I'll simply refresh it and retry. We'll see how that goes :)
I'll be pushing new version later this day.

**Reply by Bukk94** [reply, marked as answer]
Ok changes are pushed and new 1.3.0 version was release. Give it a go :) @Robertsmania

**Reply by Robertsmania** [reply]
I got the new KickLib (0.1.3) from NuGet and indeed the XSRF message does not display, but sadly when I try to send subsequent messages they do not get sent.  Here is an example to demonstrate:

```
/*
var userName = "channelUsername";
string sekritPW = "password";
string sekritEmail = "email";
Int32 channelID = 123456;
*/

IKickApi kickApi = new KickApi();

// Get information about user
var user = await kickApi.Users.GetUserAsync(userName);

// Get information about channel
var channelInfo = await kickApi.Channels.GetChannelInfoAsync(userName);

// Gets detailed information about current livestream
var liveInfo = await kickApi.Livestream.GetLivestreamInfoAsync(userName);

// Get channel clips
var channelClips = await kickApi.Clips.GetChannelClipsAsync(userName);

// Authenticated calls
//await kickApi.AuthenticateAsync("username", "password");
await kickApi.AuthenticateAsync(sekritEmail, sekritPW);
await kickApi.Messages.SendMessageAsync(channelID, "My message");

IKickClient client = new KickClient();

client.OnMessage += delegate (object sender, ChatMessageEventArgs e)
{
    Console.WriteLine(e.Data.Sender.Username + " said: " + e.Data.Content);
};

//await client.ListenToChatRoomAsync(123456);
await client.ListenToChatRoomAsync(channelID);
await client.ConnectAsync();

while (true)
{
    await Task.Delay(10000);
    string currentTime = DateTime.Now.ToString("hh:mm:ss tt");
    Console.WriteLine(currentTime);  
    await kickApi.Messages.SendMessageAsync(channelID, $"Now - {currentTime}");
}
```

The initial "My message" is sent and appears in the Kick chat.
The first "Now - time" is also sent and appears in the chat, as well as the console output since its listening by then.
But the subsequent timed messages do not show up in the chat - the ConsoleWriteLine shows when they are attempted but neither the console output nor the Kick chat itself shows those additional messages.

It is odd, in that if I set a breakpont on the await kickApi.Messages.SendMessgesAsync() call and manually step, it does send (sometimes).  

So I'm not sure if there is a timing issue or something, but waiting 10 seconds between attempts seems conservative and I don't think we're hitting a rate limit or anything?

**Reply by Bukk94** [reply]
That's weird. Is this happening only with new version or this behavior was present in older version as well?
I don't think we are hitting rate limit as there are just a few requests, but current private API is behind proxy / Cloudflare (or similar DDOS protection). That's what makes this so difficult. Instead of getting a token and using the API we need to maneuver around the protection. They said there will be official API "soon". 3 months ago :D 

Anyway, back to the problem. So it won't send the message but won't throw anything?

EDIT: I just tested it and it seems to work. Weird. I wonder if there are some limits indeed.

**Reply by Robertsmania** [reply]
Hmm, okay - you tried it with the same sort of loop test above?  I went back to 0.1.2 and see consistent behavior there - the timed messages are not showing up but I discovered that other chat activity does get the last one to go through.

Its a little hard to describe, so here's a visual aid to show what I'm seeing:
https://www.youtube.com/watch?v=oyKbU8f02Yg

**Reply by Bukk94** [reply]
Thanks for the video, it really helped - indeed I was able to reproduce the problem. After tinkering with it for a while, it seems that this is some kind of Bot Protection on Kick side. When I used your code, it did the same thing - send message and nothing happened (btw it returns 200 as 'sent' but nothing actually happens - probably due to bot protection).
But if you try to create some dictionary of random sentences and try to post those instead of time, it works. At least it own in my setup, it send 6 messages in 1 sec interval.

**Reply by Robertsmania** [reply]
Thats very strange, but okay.  I'll keep working on my project and see if the messages it needs to send are human-like enough to get sent!

**Reply by Yunusolcay** [reply]
The system does not send messages. Even though F2A and user information are correct, I am having problems. What could be the problem?

https://www.youtube.com/watch?v=FBKVeBYiXQ0

```c#
sendMessage sd = new sendMessage();
int chatRoomID = 25099174;

IKickClient client = new KickClient();
#pragma warning disable CS8622 // Nullability of reference types in type of parameter doesn't match the target delegate (possibly because of nullability attributes).
client.OnMessage += async delegate (object sender, ChatMessageEventArgs e)
{
    Console.WriteLine(e.Data.Sender.Username + " : " + e.Data.Content);
    await sd.sendAsync(chatRoomID);
};
#pragma warning restore CS8622 // Nullability of reference types in type of parameter doesn't match the target delegate (possibly because of nullability attributes).
// 25099174
await client.ListenToChatRoomAsync(chatRoomID);
await client.ConnectAsync();
Console.ReadLine();


public async Task sendAsync(int chatroomID)
{
    IKickApi kickApi = new KickApi();
    var authSettings = new AuthenticationSettings(userName, password)
    {
        TwoFactorAuthCode = twoFactor
    };
    await kickApi.AuthenticateAsync(authSettings);
    await Task.Delay(1000);
    await kickApi.Messages.SendMessageAsync(chatroomID, "deneme");
}
``` 



``

**Reply by Bukk94** [reply]
@Yunusolcay I tested the code provided in README example and it works.
Your code has few issues:
First of all, `new KickApi()` and `await kickApi.AuthenticateAsync(authSettings);` should be performed only once, ideally during startup, not every time you are sending the message. Also the delay between authenticate and send message is not really necessary.

Authentication (calling `await kickApi.AuthenticateAsync(authSettings);`) is slow due to Kick protection and can take 20-30 seconds. That's probably your problem, your program probably ends sooner than it evaluates. To prevent this, do step above -> initialize API and authenticate on your application startup, then sending messages should be instantaneous. 

Lastly, I think you might have some problems with OnMessage async delegate. The event handler is not meant to be async so making it `async void` you can't properly await it might misbehave.

**Comment by Yunusolcay** [comment]
@Bukk94 
await kickApi.AuthenticateAsync(authSettings);
When this line runs, the program freezes. It doesn't work when I put it in try catch.

The information should be like this, right?
userName email, password kick password

```c#
int chatRoomID = 25099174;

IKickApi kickApi = new KickApi();
var authSettings = new AuthenticationSettings(userName, password)
{
    TwoFactorAuthCode = twoFactor
};
await kickApi.AuthenticateAsync(authSettings);
await kickApi.Messages.SendMessageAsync(chatRoomID, "bot bağlandı");

IKickClient client = new KickClient();

#pragma warning disable CS8622 // Nullability of reference types in type of parameter doesn't match the target delegate (possibly because of nullability attributes).
client.OnMessage += delegate (object sender, ChatMessageEventArgs e)
{
    Console.WriteLine(e.Data.Sender.Username + " : " + e.Data.Content);
    kickApi.Messages.SendMessageAsync(chatRoomID, "deneme");
};
#pragma warning restore CS8622 // Nullability of reference types in type of parameter doesn't match the target delegate (possibly because of nullability attributes).
await client.ListenToChatRoomAsync(chatRoomID);
await client.ConnectAsync();
Console.ReadLine();
```

**Reply by Bukk94** [reply]
If your code is running in UI Thread then yes, whole program freezes. Wait 20-30 seconds, it should get pass that. To avoid any UI freezes this should be run somewhere in the background.
To properly await the SendMessageAsync, you can change that to:
`kickApi.Messages.SendMessageAsync(chatRoomID, "deneme").GetAwaiter().GetResults();`
Otherwise all remaining setup seems correct. It just takes a lot of time to authenticate.

**Reply by Yunusolcay** [reply]
await kickApi.AuthenticateAsync(authSettings);
use await : 
Even though I waited for 5 minutes, there was no improvement.
does not move to the bottom line
not use await :
Cannot send authenticated request without authenticating first! Call 'AuthenticateAsync' first

**Reply by Bukk94** [reply]
Yes, you must use `await`. What comes to my mind is that maybe your internet connection is slow? KickLib is using Puppeteer on the backend to create requests. It's using headless Chrome. So first time running it, it actually downloads whole headless chrome (~400 MB). You should see folder `.local-chromium` in your folder where .exe is executed.

**Reply by Yunusolcay** [reply]
There is no problem when I just want to list the messages. There is a local-chromium folder with a size of 417mb.
